packages/open-truss/src/uqi/uqi.ts

```diff
@@ -0,0 +1,176 @@
+  ) => Promise<AsyncIterableIterator<UqiResult>>
+  teardown: (context: UqiContext<C, T>) => Promise<void>
+}
+
+export interface UqiContext<C, T> {
+  config: C
+  client: T
+  status: UqiStatus
+  typeMappings: UqiTypeMappings
+}
+
+export interface UqiResult {
+  row: UqiScalar[]
+  metadata: {
+    columns: UqiColumn[]
+  }
+}
+
+export interface UqiStatus {
+  completedAt: Date | null
+  failedAt: Date | null
+  failedReason: string | null
+  recordsReturned: number
+  startedAt: Date | null
+}
+
+export default function uqi<C, T>(og: UqiSettings<C, T>): UqiClient {
+  const context: UqiContext<C, T> = {
+    config: og.config,
+    client: og.client,
+    status: {
+      completedAt: null,
+      failedAt: null,
+      failedReason: null,
+      recordsReturned: 0,
+      startedAt: null,
+    },
+    typeMappings: og.typeMappings,
+  }
+
+  function buildRow(result: UqiResult): UqiScalar[] {
+    return result.metadata.columns.map((column: UqiColumn, i: number) => {
+      const type = context.typeMappings[column.type]
+      if (!type) {
+        throw new Error(`Type ${column.type} is not mapped`)
+      }
+      const value = result.row[i]
+      if (value === null) {
+        return null
+      }
+      if (typeof value === 'string' && type === 'Number') {
+        return Number(value)
+      }
+      if (typeof value === 'number' && type === 'String') {
+        return String(value)
+      }
+      if (typeof value === 'bigint' && type === 'BigInt') {
+        return BigInt(value)
+      }
+      if (typeof value === 'string' && type === 'Date') {
+        return new Date(value)
+      }
+      return value
+    })
+  }
+
+  function buildMetadata(metadata: { columns: UqiColumn[] }): {
+    columns: UqiColumn[]
+  } {
+    return {
+      ...metadata,
+      columns: metadata.columns.map((column: UqiColumn) => {
+        const type = context.typeMappings[column.type]
+        if (!type) {
+          throw new Error(`Type ${column.type} is not mapped`)
+        }
+        return {
+          ...column,
+          type,
+        }
+      }),
+    }
+  }
+
+  return {
+    async query(
+      query: string,
+      options?: { statusCallback?: (status: UqiStatus) => Promise<void> },
+    ) {
+      if (!context.client) {
+        context.status.failedReason = 'Client is not set up'
+        context.status.failedAt = new Date()
+        if (options?.statusCallback) {
+          await options.statusCallback(context.status)
+        }
+        throw new Error('Client is not set up')
+      }
+      const queryIterator = await og.query(context, query)
+
+      context.status.startedAt = new Date()
+      if (options?.statusCallback) {
+        await options.statusCallback(context.status)
+      }
+
+      async function* asyncGenerator(): AsyncGenerator<UqiResult> {
+        try {
+          for await (const result of queryIterator) {
+            const row = buildRow(result)
+            const metadata = buildMetadata(result.metadata)
```

Comment:
I assume the answer is yes, but will there be cases where each row has different metadata? Kind of unfortunate to have to recalculate this every time.

Reply:
Fwiw I wasn't going to start optimizing performance on any parts of this (besides using the iterator/generator pattern) until we needed to. That said I think only calculating metadata when needed makes sense.

I think we have two options:
1. Update the API to return a function instead of the already calculated metadata so that you can optionally build the metadata when you need it.
2. Cache or memoize the calculated metadata.

Preference?

Reply:
number 2 makes sense to me. can probably cache it based on the metadata object. but also 👍 for optimizing later.